applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h

```diff
@@ -142,42 +142,51 @@ class MapperVertexMorphingMatrixFree : public Mapper
         mValuesDestination[2].clear();
 
         // Perform mapping
-        for(auto& node_i : mrDestinationModelPart.Nodes())
+        #pragma omp parallel for
+        for(int node_itr=0; node_itr < static_cast<int>(mrDestinationModelPart.NumberOfNodes()); node_itr++)
         {
+            auto node_i = mrDestinationModelPart.NodesBegin() + node_itr;
+
             NodeVector neighbor_nodes(mMaxNumberOfNeighbors);
             std::vector<double> resulting_squared_distances(mMaxNumberOfNeighbors);
-            unsigned int number_of_neighbors = mpSearchTree->SearchInRadius( node_i,
+            unsigned int number_of_neighbors = mpSearchTree->SearchInRadius( *node_i,
                                                                              mFilterRadius,
                                                                              neighbor_nodes.begin(),
                                                                              resulting_squared_distances.begin(),
                                                                              mMaxNumberOfNeighbors );
 
-            ThrowWarningIfNumberOfNeighborsExceedsLimit(node_i, number_of_neighbors);
+            ThrowWarningIfNumberOfNeighborsExceedsLimit(*node_i, number_of_neighbors);
 
             std::vector<double> list_of_weights( number_of_neighbors, 0.0 );
             double sum_of_weights = 0.0;
-            ComputeWeightForAllNeighbors( node_i, neighbor_nodes, number_of_neighbors, list_of_weights, sum_of_weights );
+            ComputeWeightForAllNeighbors( *node_i, neighbor_nodes, number_of_neighbors, list_of_weights, sum_of_weights );
 
-            int node_i_mapping_id = node_i.GetValue(MAPPING_ID);
+            int node_i_mapping_id = node_i->GetValue(MAPPING_ID);
             for(unsigned int neighbor_itr = 0 ; neighbor_itr<number_of_neighbors ; neighbor_itr++)
             {
                 double weight = list_of_weights[neighbor_itr] / sum_of_weights;
 
                 ModelPart::NodeType& node_j = *neighbor_nodes[neighbor_itr];
                 array_3d& nodal_variable = node_j.FastGetSolutionStepValue(rOriginVariable);
 
+                #pragma omp atomic
                 mValuesDestination[0][node_i_mapping_id] += weight*nodal_variable[0];
+                #pragma omp atomic
                 mValuesDestination[1][node_i_mapping_id] += weight*nodal_variable[1];
+                #pragma omp atomic
                 mValuesDestination[2][node_i_mapping_id] += weight*nodal_variable[2];
```

Comment:
can you reformulate those statements to only have one atomic?

Reply:
Why? The only option I see here is to have a for loop with 3 iterations and the atomic command inside the loop. But I suppose this is not much faster, is it? Or is there a way to apply the atomic keyword to a whole block?

Reply:
I think it would be more efficient to have a "vector" version that does it all at once, but I am not sure if it is possible in your case

You can also leave it as is